### ISSUE TITLE
[Feature Request] Adaptive-K Routing: Dynamic Expert Selection Based on Router Entropy

### ISSUE THREAD

**Issue by Gabrobals** [NVIDIA-NeMo/Nemotron#58]
## Summary

I've validated **Adaptive-K routing** on Nemotron 3 Nano, achieving **33.3% compute savings** by dynamically selecting K based on router entropy. This is a drop-in optimization requiring no retraining.

## Validation Results (January 19, 2026)

**Model**: nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16

| Metric | Baseline | Adaptive-K | 
|--------|----------|------------|
| Average K | 6.0 | 4.0 |
| Expert Compute | 100% | 66.7% |
| **Savings** | - | **33.3%** |

Router entropy averages 5.23 bits out of 7.0 max (log₂(128)), indicating the router is often confident enough that 6 experts is unnecessary.

## Proposed Algorithm

```python
router_logits = hidden_states @ router.weight.T
probs = softmax(router_logits)
entropy = -sum(probs * log(probs))

# Entropy-based K selection
if entropy < 4.5:
    K = 2  # High confidence
elif entropy < 5.5:
    K = 4  # Moderate
else:
    K = 6  # Baseline (uncertain)

indices, weights = top_k(probs, K)

Why This Fits Nemotron 3
Shared expert safety net: Always-active shared expert maintains quality at low K
Complements reasoning budget control: Automates per-token compute optimization
Zero retraining: Inference-time only modification
Methodology
Since output_router_logits isn't supported, I used forward hooks on backbone.layers.X.mixer.gate modules to extract pre-top-k logits.

Open Source
Repository: https://github.com/Gabrobals/sbm-efficient
Validation data: results/nemotron3_nano_validation.json
Validation script: scripts/nemotron3_entropy_validation.py
Request
Would the team be interested in:

Integrating Adaptive-K as an optional routing mode in the cookbooks?
Exposing output_router_logits for easier entropy analysis?
Happy to contribute a PR!
